### PR TITLE
lakectl autocomplete with repository name

### PR DESCRIPTION
**Pull request by nopcoder**
Enhance the auto-complete with auto complete by listing the current repositories.
This is the first step in enable better user-experience while working from the command line.
Next step will be to enable specific tag/branch/commit and path.

Part of #654

### CODE REVIEW

**Review by johnnyaug — CHANGES_REQUESTED** [review]
1. Prefix extraction seems broken: I think all repositories are always fetchedc.
2. Couldn't get it to work on my machine (Mac + Zsh).

File: cmd/lakectl/cmd/validargs.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	// do not suggest in case we are passed the repository part
+	uriPrefix := uri.LakeFSSchema + uri.LakeFSSchemaSeparator
+	if strings.HasPrefix(toComplete, uriPrefix) && strings.Contains(toComplete[len(uriPrefix):], uri.PathSeparator) {
+		return nil, cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveNoSpace
+	}
+
+	// extract repository name written so far
+	var prefix api.PaginationPrefix
+	if strings.HasPrefix(toComplete, uriPrefix) {
+		idx := strings.Index(toComplete[len(uriPrefix):], uri.PathSeparator)
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/validargs.go:29]: True, fixing.

**Comment by nopcoder** [comment]
> 1. Prefix extraction seems broken: I think all repositories are always fetched.
> 2. Couldn't get it to work on my machine (Mac + Zsh).

Thanks for the review - fixing the prefix use. if you like to get it working on your mac, try: `lakectl completion zsh > "${fpath[1]}/_lakectl"`

**Comment by johnnyaug** [comment]
@nopcoder worked! Thanks

**Comment by nopcoder** [comment]
> @nopcoder worked! Thanks

Pushed the fix - now the repository prefix passed to to the api call.

File: cmd/lakectl/cmd/validargs.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	uriPrefix := uri.LakeFSSchema + uri.LakeFSSchemaSeparator
+	if strings.HasPrefix(toComplete, uriPrefix) && strings.Contains(toComplete[len(uriPrefix):], uri.PathSeparator) {
+		return nil, cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveNoSpace
+	}
+
+	// extract repository name written so far
+	var prefix api.PaginationPrefix
+	if strings.HasPrefix(toComplete, uriPrefix) {
+		if !strings.Contains(toComplete[len(uriPrefix):], uri.PathSeparator) {
+			prefix = api.PaginationPrefix(toComplete[len(uriPrefix):])
```
Comment (johnnyaug) [inline cmd/lakectl/cmd/validargs.go:30]: else - prefix = toComplete?

File: cmd/lakectl/cmd/validargs.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+func validRepositoryToComplete(ctx context.Context, toComplete string) ([]string, cobra.ShellCompDirective) {
+	// do not suggest in case we are passed the repository part
+	uriPrefix := uri.LakeFSSchema + uri.LakeFSSchemaSeparator
+	if strings.HasPrefix(toComplete, uriPrefix) && strings.Contains(toComplete[len(uriPrefix):], uri.PathSeparator) {
+		return nil, cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveNoSpace
+	}
+
+	// extract repository name written so far
+	var prefix api.PaginationPrefix
+	if strings.HasPrefix(toComplete, uriPrefix) {
```
Comment (johnnyaug) [inline cmd/lakectl/cmd/validargs.go:28]: If it doesn't have the `lakefs://` prefix, we still want an autocomplete, no?

So maybe the prefix should be "toComplete".
For example, if I'm doing `lakectl branch create my-rep<TAB>`, I want to get: `lakectl branch create lakefs://my-repo`

**Review by johnnyaug — APPROVED** [review]
Nice feature!! See another comment

File: cmd/lakectl/cmd/validargs.go (hunk 3 above)
Comment (nopcoder) [reply cmd/lakectl/cmd/validargs.go:28]: Wanted to give the same autocomplete behaviour we have with other application - when the prefix you enter doesn't match any of the possible options we do not suggest completion.
Example: docker autocompete will match the prefix to the image name - but if you will find me a other working reference I will consider :)

**Pull request merged**